### PR TITLE
Add fingerprint to assets (cf: CIP-0014)

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

ADP-703

# Overview

Motivation, design and rationale explained in: [CIP-0014](https://github.com/cardano-foundation/CIPs/pull/64)

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs [hunk 1]
```diff
@@ -128,6 +146,46 @@ instance FromText TokenName where
...
+    & UnsafeTokenFingerprint
+
+tokenFingerprintHrp :: Bech32.HumanReadablePart
+tokenFingerprintHrp = [humanReadablePart|asset|]
+
+instance ToText TokenFingerprint where
+    toText = unTokenFingerprint
+
+instance FromText TokenFingerprint where
+    fromText txt = case Bech32.decodeLenient txt of
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs:174]: We could add a `textRoundtrip` test I take?

**Review by Anviking — APPROVED** [review]
👌looks neat

**Comment by Anviking** [comment]
That said, it's arguably a bit complicated that:
- assetName = not per se a presentable name
- asset**Id** = hex(policyId) <> hex(assetName)   (needed in metadata, but not in wallet api)
- asset**Fingerprint** = bech32(someHash(policyId <> assetName))

File: lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs (hunk 1 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/Types/TokenPolicy.hs:174]: Added

**Comment by KtorZ** [comment]
> That said, it's arguably a bit complicated that:

Isn't it? But there isn't much we can do about it unfortunately.

**Comment by KtorZ** [comment]
bors merge

**Comment by KtorZ** [comment]
bors retry

**Pull request merged**